In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

* we need pyarrow to workout with parquet
* pip install pyarrow

In [2]:
data = pd.read_parquet('green_tripdata_2021-01.parquet')

In [3]:
data.head()

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0         2  2021-01-01 00:15:56   2021-01-01 00:19:52                  N   
1         2  2021-01-01 00:25:59   2021-01-01 00:34:44                  N   
2         2  2021-01-01 00:45:57   2021-01-01 00:51:55                  N   
3         2  2020-12-31 23:57:51   2021-01-01 00:04:56                  N   
4         2  2021-01-01 00:16:36   2021-01-01 00:16:40                  N   

   RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
0         1.0            43           151              1.0           1.01   
1         1.0           166           239              1.0           2.53   
2         1.0            41            42              1.0           1.12   
3         1.0           168            75              1.0           1.99   
4         2.0           265           265              3.0           0.00   

   fare_amount  extra  mta_tax  tip_amount  tolls_amount ehail_fee  \
0          5.5    0.5      0.5        0.00           0.0      None   
1         10.0    0.5      0.5        2.81           0.0      None   
2          6.0    0.5      0.5        1.00           0.0      None   
3          8.0    0.5      0.5        0.00           0.0      None   
4        -52.0    0.0     -0.5        0.00           0.0      None   

   improvement_surcharge  total_amount  payment_type  trip_type  \
0                    0.3          6.80           2.0        1.0   
1                    0.3         16.86           1.0        1.0   
2                    0.3          8.30           1.0        1.0   
3                    0.3          9.30           2.0        1.0   
4                   -0.3        -52.80           3.0        1.0   

   congestion_surcharge  
0                  0.00  
1                  2.75  
2                  0.00  
3                  0.00  
4                  0.00

In [4]:
import sklearn

In [5]:
sklearn.__version__

'1.5.0'

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76518 entries, 0 to 76517
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   VendorID               76518 non-null  int64         
 1   lpep_pickup_datetime   76518 non-null  datetime64[us]
 2   lpep_dropoff_datetime  76518 non-null  datetime64[us]
 3   store_and_fwd_flag     40471 non-null  object        
 4   RatecodeID             40471 non-null  float64       
 5   PULocationID           76518 non-null  int64         
 6   DOLocationID           76518 non-null  int64         
 7   passenger_count        40471 non-null  float64       
 8   trip_distance          76518 non-null  float64       
 9   fare_amount            76518 non-null  float64       
 10  extra                  76518 non-null  float64       
 11  mta_tax                76518 non-null  float64       
 12  tip_amount             76518 non-null  float64       
 13  t

In [7]:
data_train = pd.read_parquet('green_tripdata_2021-01.parquet')
data_val = pd.read_parquet('green_tripdata_2021-02.parquet')

In [8]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76518 entries, 0 to 76517
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   VendorID               76518 non-null  int64         
 1   lpep_pickup_datetime   76518 non-null  datetime64[us]
 2   lpep_dropoff_datetime  76518 non-null  datetime64[us]
 3   store_and_fwd_flag     40471 non-null  object        
 4   RatecodeID             40471 non-null  float64       
 5   PULocationID           76518 non-null  int64         
 6   DOLocationID           76518 non-null  int64         
 7   passenger_count        40471 non-null  float64       
 8   trip_distance          76518 non-null  float64       
 9   fare_amount            76518 non-null  float64       
 10  extra                  76518 non-null  float64       
 11  mta_tax                76518 non-null  float64       
 12  tip_amount             76518 non-null  float64       
 13  t

In [9]:
data.dtypes

VendorID                          int64
lpep_pickup_datetime     datetime64[us]
lpep_dropoff_datetime    datetime64[us]
store_and_fwd_flag               object
RatecodeID                      float64
PULocationID                      int64
DOLocationID                      int64
passenger_count                 float64
trip_distance                   float64
fare_amount                     float64
extra                           float64
mta_tax                         float64
tip_amount                      float64
tolls_amount                    float64
ehail_fee                        object
improvement_surcharge           float64
total_amount                    float64
payment_type                    float64
trip_type                       float64
congestion_surcharge            float64
dtype: object

In [10]:
#### Text Manipulation for reading Flexibility

In [11]:
data_train['lpep_dropoff_datetime'] = pd.to_datetime(data_train['lpep_dropoff_datetime'])
data_train['lpep_pickup_datetime'] = pd.to_datetime(data_train['lpep_pickup_datetime'])

data_val['lpep_dropoff_datetime'] = pd.to_datetime(data_val['lpep_dropoff_datetime'])
data_val['lpep_pickup_datetime'] = pd.to_datetime(data_val['lpep_pickup_datetime'])

In [12]:
data_train['duration'] = data_train['lpep_dropoff_datetime'] - data_train['lpep_pickup_datetime']

data_val['duration'] = data_val['lpep_dropoff_datetime'] - data_val['lpep_pickup_datetime']

In [13]:
data_train['duration']

0       0 days 00:03:56
1       0 days 00:08:45
2       0 days 00:05:58
3       0 days 00:07:05
4       0 days 00:00:04
              ...      
76513   0 days 00:38:00
76514   0 days 00:38:00
76515   0 days 00:11:00
76516   0 days 00:27:00
76517   0 days 00:10:00
Name: duration, Length: 76518, dtype: timedelta64[us]

In [14]:
data_train['duration'] = data_train['duration'].apply(lambda x:x.total_seconds()/60)

data_val['duration'] = data_val['duration'].apply(lambda x:x.total_seconds()/60)

In [15]:
data_train = data_train[(data_train['duration'] >= 1) & (data_train['duration'] <= 60)]

data_val = data_val[(data_val['duration'] >= 1) & (data_val['duration'] <= 60)]

In [16]:
cat = ['PULocationID' , 'DOLocationID']
num = ['trip_distance']

In [17]:
data_train[cat] = data_train[cat].astype(str)

data_val[cat] = data_val[cat].astype(str)

In [18]:
data_train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 73908 entries, 0 to 76517
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   VendorID               73908 non-null  int64         
 1   lpep_pickup_datetime   73908 non-null  datetime64[us]
 2   lpep_dropoff_datetime  73908 non-null  datetime64[us]
 3   store_and_fwd_flag     38175 non-null  object        
 4   RatecodeID             38175 non-null  float64       
 5   PULocationID           73908 non-null  object        
 6   DOLocationID           73908 non-null  object        
 7   passenger_count        38175 non-null  float64       
 8   trip_distance          73908 non-null  float64       
 9   fare_amount            73908 non-null  float64       
 10  extra                  73908 non-null  float64       
 11  mta_tax                73908 non-null  float64       
 12  tip_amount             73908 non-null  float64       
 13  tolls_

In [19]:
data_val

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0             2  2021-02-01 00:34:03   2021-02-01 00:51:58                  N   
1             2  2021-02-01 00:04:00   2021-02-01 00:10:30                  N   
2             2  2021-02-01 00:18:51   2021-02-01 00:34:06                  N   
3             2  2021-02-01 00:53:27   2021-02-01 01:11:41                  N   
4             2  2021-02-01 00:57:46   2021-02-01 01:06:44                  N   
...         ...                  ...                   ...                ...   
64567         2  2021-02-28 22:19:00   2021-02-28 22:29:00               None   
64568         2  2021-02-28 23:18:00   2021-02-28 23:27:00               None   
64569         2  2021-02-28 23:44:00   2021-02-28 23:58:00               None   
64570         2  2021-02-28 23:07:00   2021-02-28 23:14:00               None   
64571         2  2021-02-28 23:43:00   2021-02-28 23:50:00               None   

       RatecodeID PULocationID DOLocationID  passenger_count  trip_distance  \
0             1.0          130          205              5.0           3.66   
1             1.0          152          244              1.0           1.10   
2             1.0          152           48              1.0           4.93   
3             1.0          152          241              1.0           6.70   
4             1.0           75           42              1.0           1.89   
...           ...          ...          ...              ...            ...   
64567         NaN          129            7              NaN           2.63   
64568         NaN          116          166              NaN           1.87   
64569         NaN           74          151              NaN           2.40   
64570         NaN           42           42              NaN           1.11   
64571         NaN           42           41              NaN           1.40   

       fare_amount  ...  mta_tax  tip_amount  tolls_amount  ehail_fee  \
0            14.00  ...      0.5       10.00           0.0       None   
1             6.50  ...      0.5        0.00           0.0       None   
2            16.50  ...      0.5        0.00           0.0       None   
3            21.00  ...      0.5        0.00           0.0       None   
4             8.50  ...      0.5        2.45           0.0       None   
...            ...  ...      ...         ...           ...        ...   
64567        10.04  ...      0.0        0.00           0.0       None   
64568         8.33  ...      0.0        1.89           0.0       None   
64569        12.61  ...      0.0        0.00           0.0       None   
64570        11.95  ...      0.0        0.00           0.0       None   
64571         8.21  ...      0.0        0.00           0.0       None   

      improvement_surcharge  total_amount  payment_type  trip_type  \
0                       0.3         25.30           1.0        1.0   
1                       0.3          7.80           2.0        1.0   
2                       0.3         20.55           2.0        1.0   
3                       0.3         22.30           2.0        1.0   
4                       0.3         12.25           1.0        1.0   
...                     ...           ...           ...        ...   
64567                   0.3         10.34           NaN        NaN   
64568                   0.3         10.52           NaN        NaN   
64569                   0.3         12.91           NaN        NaN   
64570                   0.3         15.00           NaN        NaN   
64571                   0.3          8.51           NaN        NaN   

       congestion_surcharge   duration  
0                      0.00  17.916667  
1                      0.00   6.500000  
2                      2.75  15.250000  
3                      0.00  18.233333  
4                      0.00   8.966667  
...                     ...        ...  
64567                   NaN  10.000000  
64568                   NaN   9.000000 

In [20]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76518 entries, 0 to 76517
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   VendorID               76518 non-null  int64         
 1   lpep_pickup_datetime   76518 non-null  datetime64[us]
 2   lpep_dropoff_datetime  76518 non-null  datetime64[us]
 3   store_and_fwd_flag     40471 non-null  object        
 4   RatecodeID             40471 non-null  float64       
 5   PULocationID           76518 non-null  int64         
 6   DOLocationID           76518 non-null  int64         
 7   passenger_count        40471 non-null  float64       
 8   trip_distance          76518 non-null  float64       
 9   fare_amount            76518 non-null  float64       
 10  extra                  76518 non-null  float64       
 11  mta_tax                76518 non-null  float64       
 12  tip_amount             76518 non-null  float64       
 13  t

In [21]:
data_train = data_train[data_train['trip_type'] == 2]

data_val = data_val[data_val['trip_type'] == 2]

In [22]:
data_train['lpep_pickup_datetime'] = pd.to_datetime(data_train['lpep_pickup_datetime'])
data_train['lpep_dropoff_datetime'] = pd.to_datetime(data_train['lpep_dropoff_datetime'])

data_val['lpep_pickup_datetime'] = pd.to_datetime(data_val['lpep_pickup_datetime'])
data_val['lpep_dropoff_datetime'] = pd.to_datetime(data_val['lpep_dropoff_datetime'])

/tmp/ipykernel_6547/625001632.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_val['lpep_pickup_datetime'] = pd.to_datetime(data_val['lpep_pickup_datetime'])
/tmp/ipykernel_6547/625001632.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_val['lpep_dropoff_datetime'] = pd.to_datetime(data_val['lpep_dropoff_datetime'])


In [23]:
data_train.head()

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
30         2  2021-01-01 00:35:29   2021-01-01 00:55:15                  N   
53         2  2021-01-01 01:54:51   2021-01-01 02:15:35                  N   
69         2  2021-01-01 02:42:49   2021-01-01 02:50:59                  N   
88         2  2021-01-01 04:52:02   2021-01-01 05:05:01                  N   
96         2  2021-01-01 05:52:43   2021-01-01 05:58:02                  N   

    RatecodeID PULocationID DOLocationID  passenger_count  trip_distance  \
30         5.0           74          247              1.0           3.64   
53         5.0           74           94              1.0           5.82   
69         5.0          136          241              1.0           0.57   
88         5.0          247           75              1.0           3.43   
96         5.0            7            7              1.0           0.65   

    fare_amount  ...  mta_tax  tip_amount  tolls_amount  ehail_fee  \
30         13.0  ...      0.0         0.0           0.0       None   
53         18.0  ...      0.0         0.0           0.0       None   
69          9.0  ...      0.0         0.0           0.0       None   
88         15.0  ...      0.0         0.0           0.0       None   
96         50.0  ...      0.0         7.0           0.0       None   

   improvement_surcharge  total_amount  payment_type  trip_type  \
30                   0.3          13.3           2.0        2.0   
53                   0.3          18.3           2.0        2.0   
69                   0.3           9.3           2.0        2.0   
88                   0.3          15.3           2.0        2.0   
96                   0.3          57.3           1.0        2.0   

    congestion_surcharge   duration  
30                   0.0  19.766667  
53                   0.0  20.733333  
69                   0.0   8.166667  
88                   0.0  12.983333  
96                   0.0   5.316667  

[5 rows x 21 columns]

In [24]:
data_val.head()

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
5           2  2021-02-01 00:33:03   2021-02-01 00:40:54                  N   
9           2  2021-02-01 02:56:55   2021-02-01 02:58:38                  N   
114         2  2021-02-01 19:01:09   2021-02-01 19:11:20                  N   
125         2  2021-02-01 22:57:12   2021-02-01 23:07:36                  N   
150         2  2021-02-02 06:31:38   2021-02-02 06:50:10                  N   

     RatecodeID PULocationID DOLocationID  passenger_count  trip_distance  \
5           5.0          197          219              1.0           3.30   
9           5.0           78           78              1.0           0.00   
114         5.0           75           74              1.0           1.80   
125         5.0          223           92              1.0           4.98   
150         5.0          223          100              1.0           5.64   

     fare_amount  ...  mta_tax  tip_amount  tolls_amount  ehail_fee  \
5           15.0  ...      0.0        0.76           0.0       None   
9           24.0  ...      0.0        0.00           0.0       None   
114         15.0  ...      0.0        0.00           0.0       None   
125         50.0  ...      0.0        0.00           0.0       None   
150         50.0  ...      0.0       13.26           0.0       None   

    improvement_surcharge  total_amount  payment_type  trip_type  \
5                     0.3         16.06           1.0        2.0   
9                     0.3         24.30           2.0        2.0   
114                   0.3         15.30           1.0        2.0   
125                   0.3         50.30           2.0        2.0   
150                   0.3         66.31           1.0        2.0   

     congestion_surcharge   duration  
5                    0.00   7.850000  
9                    0.00   1.716667  
114                  0.00  10.183333  
125                  0.00  10.400000  
150                  2.75  18.533333  

[5 rows x 21 columns]

### Module : 1 -- Model Training

In [25]:
data_train['PU-DO'] = data_train['PULocationID'] + '-' + data_train['DOLocationID']

data_val['PU-DO'] = data_val['PULocationID'] + '-' + data_val['DOLocationID']

/tmp/ipykernel_6547/721656174.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_val['PU-DO'] = data_val['PULocationID'] + '-' + data_val['DOLocationID']


In [26]:
catg = ['PU-DO']
nume = ['trip_distance']

In [27]:
from sklearn.feature_extraction import DictVectorizer

In [28]:
dv = DictVectorizer()

In [29]:
train_dict = data_train[catg +  nume].to_dict(orient='records')
X_train = dv.fit_transform(train_dict)

val_dict = data_val[catg +  nume].to_dict(orient='records')
X_val = dv.transform(val_dict)

In [30]:
target = 'duration'

In [31]:
Y_train = data_train[target].values
Y_val = data_val[target].values

#### Wroking of Dict Vectorizer

In [32]:
from sklearn.feature_extraction import DictVectorizer

# Example data: list of dictionaries where each dictionary represents a sample
data = [
    {'city': 'New York', 'temperature': 32},
    {'city': 'San Francisco', 'temperature': 21},
    {'city': 'Chicago', 'temperature': 14},
    {'city': 'Boston', 'temperature': 29}
]

# Create an instance of DictVectorizer
vectorizer = DictVectorizer(sparse=False)  # sparse=False makes output dense, not sparse matrix

# Fit and transform the data with DictVectorizer
X = vectorizer.fit_transform(data)

# Print the feature names derived from the dictionaries
print("Feature names:", vectorizer.get_feature_names_out())

# Print the transformed data (numerical feature matrix)
print("Transformed data:\n", X)


Feature names: ['city=Boston' 'city=Chicago' 'city=New York' 'city=San Francisco'
 'temperature']
Transformed data:
 [[ 0.  0.  1.  0. 32.]
 [ 0.  0.  0.  1. 21.]
 [ 0.  1.  0.  0. 14.]
 [ 1.  0.  0.  0. 29.]]


#### Label Encoder Function

In [33]:
from sklearn.preprocessing import LabelEncoder

# Example data: categorical labels
labels = ['apple', 'banana', 'cherry', 'apple', 'banana']

# Create an instance of LabelEncoder
encoder = LabelEncoder()

# Fit and transform the labels
encoded_labels = encoder.fit_transform(labels)

# Print the encoded labels
print("Encoded labels:", encoded_labels)

# Reverse transformation (if needed)
decoded_labels = encoder.inverse_transform(encoded_labels)
print("Decoded labels:", decoded_labels)


Encoded labels: [0 1 2 0 1]
Decoded labels: ['apple' 'banana' 'cherry' 'apple' 'banana']


In [34]:
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.metrics import mean_squared_error

In [35]:
lr = LinearRegression()
ridg = Ridge()
lass = Lasso()

In [36]:
lr.fit(X_train , Y_train)
ridg.fit(X_train , Y_train)
lass.fit(X_train , Y_train)

Lasso()

In [37]:
Y_pred_lr = lr.predict(X_val)
Y_pred_ridg = ridg.predict(X_val)
Y_pred_lass = lass.predict(X_val)

In [38]:
print('Linear Regression MSE : ' , mean_squared_error(Y_val , Y_pred_lr , squared=False))
print('Ridge Regression MSE : ' , mean_squared_error(Y_val , Y_pred_ridg , squared=False))
print('Lasso Regression MSE : ' , mean_squared_error(Y_val , Y_pred_lass , squared=False))

Linear Regression MSE :  8.353710401133316
Ridge Regression MSE :  8.181513722504025
Lasso Regression MSE :  8.672902008635951


/home/codespace/.local/lib/python3.10/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(
/home/codespace/.local/lib/python3.10/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(
/home/codespace/.local/lib/python3.10/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


#### Ridge and Lasso with lr 

In [39]:
ridg_lr = Ridge(alpha= 0.00101)
lass_lr = Lasso(alpha=0.00101)

In [40]:
ridg_lr.fit(X_train , Y_train)
lass_lr.fit(X_train , Y_train)

Lasso(alpha=0.00101)

In [41]:
Y_pred_ridg_lr = ridg.predict(X_val)
Y_pred_lass_lr = lass.predict(X_val)

In [42]:
print('Ridge Regression with lr = 0.01 MSE : ' , mean_squared_error(Y_val , Y_pred_ridg_lr , squared=False))
print('Lasso Regression with lr = 0.01 MSE : ' , mean_squared_error(Y_val , Y_pred_lass_lr , squared=False))

Ridge Regression with lr = 0.01 MSE :  8.181513722504025
Lasso Regression with lr = 0.01 MSE :  8.672902008635951


/home/codespace/.local/lib/python3.10/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(
/home/codespace/.local/lib/python3.10/site-packages/sklearn/metrics/_regression.py:492: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


#### MSE Values

* Linear Regression MSE :  8.353710401133316
* Ridge Regression MSE :  8.181513722504025
* Lasso Regression MSE :  8.672902008635951

**with Alpha = 0.01**

* Ridge Regression with lr = 0.01 MSE :  8.181513722504025
* Lasso Regression with lr = 0.01 MSE :  8.672902008635951